packages/x-data-grid-pro/src/components/GridServerSideTreeDataGroupingCell.tsx

```diff
@@ -0,0 +1,220 @@
+  hideDescendantCount?: boolean;
+  /**
+   * The cell offset multiplier used for calculating cell offset (`rowNode.depth * offsetMultiplier` px).
+   * @default 2
+   */
+  offsetMultiplier?: number;
+}
+
+interface GridTreeDataGroupingCellIconProps
+  extends Pick<GridTreeDataGroupingCellProps, 'id' | 'field' | 'rowNode' | 'row'> {
+  descendantCount: number;
+}
+
+const LoadingContainer = styled('div')({
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'center',
+  height: '100%',
+});
+
+function GridTreeDataGroupingCellIcon(props: GridTreeDataGroupingCellIconProps) {
+  const apiRef = useGridPrivateApiContext() as React.MutableRefObject<GridPrivateApiPro>;
+  const rootProps = useGridRootProps();
+  const { rowNode, id, field, descendantCount } = props;
+
+  const loadingSelector = React.useCallback(
+    (state: GridStatePro) => state.serverSideData.loading[id] ?? false,
+    [id],
+  );
+  const errorSelector = React.useCallback(
+    (state: GridStatePro) => state.serverSideData.errors[id] ?? null,
+    [id],
+  );
+  const isDataLoading = useGridSelector(apiRef, loadingSelector);
+  const error = useGridSelector(apiRef, errorSelector);
+
+  const isServerSideNode = rowNode.isServerSide;
+
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    if (!rowNode.childrenExpanded) {
+      // always fetch/get from cache the children when the node is expanded
+      apiRef.current.enqueueChildrenFetch(id);
+    } else {
+      apiRef.current.setRowChildrenExpansion(id, !rowNode.childrenExpanded);
+    }
+    apiRef.current.setCellFocus(id, field);
+    event.stopPropagation(); // TODO remove event.stopPropagation
+  };
+
+  const Icon = rowNode.childrenExpanded
+    ? rootProps.slots.treeDataCollapseIcon
+    : rootProps.slots.treeDataExpandIcon;
+
+  if (isDataLoading) {
+    return (
+      <LoadingContainer>
+        <CircularProgress size="1rem" color="inherit" />
+      </LoadingContainer>
+    );
+  }
+  return descendantCount > 0 || isServerSideNode ? (
+    <rootProps.slots.baseIconButton
+      size="small"
+      onClick={handleClick}
+      tabIndex={-1}
+      aria-label={
+        rowNode.childrenExpanded
+          ? apiRef.current.getLocaleText('treeDataCollapse')
+          : apiRef.current.getLocaleText('treeDataExpand')
+      }
+      {...rootProps?.slotProps?.baseIconButton}
+    >
+      <rootProps.slots.baseTooltip title={error?.message ?? null}>
+        <Badge variant="dot" color="error" invisible={!error}>
+          <Icon fontSize="inherit" />
+        </Badge>
+      </rootProps.slots.baseTooltip>
+    </rootProps.slots.baseIconButton>
+  ) : null;
+}
+
+function GridServerSideTreeDataGroupingCell(props: GridTreeDataGroupingCellProps) {
+  const { id, field, formattedValue, rowNode, hideDescendantCount, offsetMultiplier = 2 } = props;
+
+  const rootProps = useGridRootProps();
+  const apiRef = useGridPrivateApiContext();
+  const row = apiRef.current.getRow(rowNode.id);
```

Comment:
Using an API method here means there is no reactive binding between the row and this component. It's going to work because this component is currently not `react.memo`'d, but we should take the habit of using hooked grid selectors to avoid bugs in the future.

Reply:
Good point, derived selectors are nice to use. Updated!